pkg/encryption/cipher.go

```diff
@@ -3,183 +3,183 @@ package encryption
 import (
 	"crypto/aes"
 	"crypto/cipher"
-	"crypto/hmac"
 	"crypto/rand"
-	"crypto/sha1"
-	"crypto/sha256"
 	"encoding/base64"
 	"fmt"
-	"hash"
 	"io"
-	"net/http"
-	"strconv"
-	"strings"
-	"time"
 )
 
-// SecretBytes attempts to base64 decode the secret, if that fails it treats the secret as binary
-func SecretBytes(secret string) []byte {
-	b, err := base64.RawURLEncoding.DecodeString(strings.TrimRight(secret, "="))
-	if err == nil {
-		// Only return decoded form if a valid AES length
-		// Don't want unintentional decoding resulting in invalid lengths confusing a user
-		// that thought they used a 16, 24, 32 length string
-		for _, i := range []int{16, 24, 32} {
-			if len(b) == i {
-				return b
-			}
-		}
-	}
-	// If decoding didn't work or resulted in non-AES compliant length,
-	// assume the raw string was the intended secret
-	return []byte(secret)
+// Cipher provides methods to encrypt and decrypt
+type Cipher interface {
+	Encrypt(value []byte) ([]byte, error)
+	Decrypt(ciphertext []byte) ([]byte, error)
+	EncryptInto(s *string) error
+	DecryptInto(s *string) error
 }
 
-// cookies are stored in a 3 part (value + timestamp + signature) to enforce that the values are as originally set.
-// additionally, the 'value' is encrypted so it's opaque to the browser
-
-// Validate ensures a cookie is properly signed
-func Validate(cookie *http.Cookie, seed string, expiration time.Duration) (value string, t time.Time, ok bool) {
-	// value, timestamp, sig
-	parts := strings.Split(cookie.Value, "|")
-	if len(parts) != 3 {
-		return
-	}
-	if checkSignature(parts[2], seed, cookie.Name, parts[0], parts[1]) {
-		ts, err := strconv.Atoi(parts[1])
-		if err != nil {
-			return
-		}
-		// The expiration timestamp set when the cookie was created
-		// isn't sent back by the browser. Hence, we check whether the
-		// creation timestamp stored in the cookie falls within the
-		// window defined by (Now()-expiration, Now()].
-		t = time.Unix(int64(ts), 0)
-		if t.After(time.Now().Add(expiration*-1)) && t.Before(time.Now().Add(time.Minute*5)) {
-			// it's a valid cookie. now get the contents
-			rawValue, err := base64.URLEncoding.DecodeString(parts[0])
-			if err == nil {
-				value = string(rawValue)
-				ok = true
-				return
-			}
-		}
-	}
-	return
+type Base64Cipher struct {
```

Comment:
Just wondering if this and the other Ciphers should be private? Since we have constructors and an interface I don't think they need to be public, WDYT?

Reply:
In the upcoming session refactor that builds off of this, different use cases use different encryption ciphers that are best for various use cases:

CFB, GCM that operate on pure `[]byte` and never have a `string` are used with the MessagePack encoding (this is the secret sauce that allows us not to have a base64 encoding adding 33% bloat to fit an encrypted value into JSON that requires a string).

For Cookie sessions, pure CFB (not base64 wrapped) is used on encrypting sessions.  CFB is missing authentication (which GCM has) but isn't weak to IV reuse issues (which prevents us from using GCM comfortably in cookies in case some users never cycle the cookie secret).  We get authentication of the cookie via the SHA256 HMAC on the cookie.

For the encryption layer in DB based session stores (currently just Redis), GCM is great because it is better in all aspects than GCM, and it doesn't have the IV reuse issue since we make a different secret that is unique to each session.

Base64 wrapped versions are left available for backwards compatibility with legacy JSON based sessions that still exist.

Hence, I left all public and accessible.

Reply:
Oh, lol, I think I follow your suggestion now 😄 

Treat the above as me over explaining a question you didn't ask.

Reply:
So what's a good naming convention for the `CFBCipher` and `GCMCipher` to make the structs themselves private?

Reply:
I would just go for `cfbCipher` and `gcmCipher` if that's ok with you?

Reply:
lol -- so obvious 🤦